### PR TITLE
Fix GPIO16 Pull-down

### PR DESCRIPTION
**Pull request by netpok**
Currently the pull-up settings are not applied on GPIO16, this pull request fixes this behavior and allows the internal pull-down to be disabled.

### CODE REVIEW

**Comment by andrethomas** [comment]
Just a side note - This pin has a special function which is why it has internal pull-down... it is used to wake up from deep sleep by the internal timer interrupt pulling the pin high so that the ESP8266 resets and starts running.

Yes, I know this is not used in Tasmota but the challenge is that in a lot of cases (ESP8266-12E, for example) the GPIO16 pin is actually hard-wired to the RESET pin.

Caution is therefore advised when using this as an input pin because it can have adverse implications on some ESP8266 based modules.

If used for a push button that pulls GPIO16 down on a module that has GPIO16 hard-wired to the RESET pin this would cause a hard reset on the ESP8266 when the button is released.

**Comment by andrethomas** [comment]
Try this from the command line console

`GPIO16 1`

It will cause an immediate reset with the restart reason being "Software/System restart" if connected to the reset pin (as seems to be the case on the ESP-07 also which I just tested it with)

**Comment by netpok** [comment]
`GPIO16 1` will cause software reset, because it sets the function of gpio 16 and then restart the device to reload the config, executing `GPIO14 1` will also cause a system restart.

I made this pull request because I have a Digoo SP202 smart socket, which uses GPIO16 as a button via an external 10k resistor, leaving the internal pull-down active causes ghost switching, because the pull-down and the pull-up resistor forms a voltage divider with a ~2.5V output.

**Comment by Jason2866** [comment]
It is a very bad idea to change it in general for all devices. Many issues will occur.

**Comment by netpok** [comment]
It's not changing it in general.

I think this was a bug:
Old behavior:

- Normal gpio configured as `KEY1`: Pull-up enabled
- Normal gpio configured as `KEY1_NP`: Pull-up disabled
- gpio16 configured as `KEY1`: Pull-down enabled
- gpio16 configured as `KEY1_NP`: Pull-down **enabled**

New, fixed behavior:

- Normal gpio configured as `KEY1`: Pull-up enabled
- Normal gpio configured as `KEY1_NP`: Pull-up disabled
- gpio16 configured as `KEY1`: Pull-down enabled
- gpio16 configured as `KEY1_NP`: Pull-down **disabled**

As `_NP` stands for no pull, I think this is the correct way to handle buttons, also in the template there are no switch or button defined on gpio16 with no pull, so it should not change anything for currently supported devices.

**Comment by andrethomas** [comment]
From the datasheet

image

GPIO pin will be triggered as HIGH with a voltage of > 0.75v - 0.80v

You would need a voltage lower than 0.75 volts to cause ghost switching.

**Comment by netpok** [comment]
Maybe I measured it wrong, but disabling the internal pull down fixed the problem. Also based on my previous comment I think this is a general bug, since `_NP` version off switches and buttons are used only if someone wants to disable the internal pull resistors.

**Comment by andrethomas** [comment]
@netpok I guess if it solves it the measurement was maybe on the wrong end of the 3.3V - i.e. 3.3V - the 2.5 you measured = 1.3V - maybe it dips below that at some point due to instabilities inside the Digoo SP202 to match a low condition on the GPIO pin. 

So all the information is here now... now it's up to @arendst whether he wants to merge or not.

If we start getting users with random boot loops we'll know why :)

**Review by andrethomas2 — APPROVED** [review]
Approving based on the code being fine - Cannot guarantee any implications but we can observe the current development codebase binary usage and see if other users report unpredictable behaviour based on this PR.

**Pull request merged**
